design/aws-ecs-service-discovery-integration.md

```diff
@@ -0,0 +1,235 @@
+# AWS ECS - Support for AWS Cloud Map (Service Discovery) Integration
+
+To address issue [#1554](https://github.com/awslabs/aws-cdk/issues/1554), the
+ECS construct library should provide a way to set
+[`serviceRegistries`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_CreateService.html#ECS-CreateService-request-serviceRegistries)
+on their ECS service at the L1 level.
+
+
+## General approach
+Rather than having the customer instantiate Cloud Map constructs directly and
+thus be forced to learn the API of a separate construct library, we should
+allow the customer to pass in the domain name and other configuration minimally
+needed to instantiate a Cloud Map namespace within an ECS cluster and create a
+Cloud Map service for each ECS service.
+
+The current proposal is to add a method on the ECS Cluster construct,
+`addNamespace` that would take a set of properties that include the namespace
+type (Public DNS, Private DNS or Http) and namespace name (domain name).
+
+While it is possible to use more than one namespace for services in an ECS
+cluster, realistically, we would expect ECS customers to only have one
+namespace per given cluster. A Cloud Map service within that namespace would
+then be created for each ECS service using service discovery in the cluster,
+and would be discoverable by service name within that namespace, e.g.
+frontend.mydomain.com, backend.mydomain.com, etc.
+
+ECS will automatically register service discovery instances that are accessible
+by IP address (IPv4 only) on the createService API call.
+
+// FIXME public namespace needs to be imported?
+
+## Code changes
+
+The following are the new methods/interfaces that would be required for the proposed approach:
+
+#### Cluster#addNamespace(options: NamespaceOptions)
+
+This will allow adding a Cloud Map namespace, which will be accessible as a
+member on the cluster. In the case of a Private DNS Namespace, a Route53 hosted
+zone will be created for the customer.
+
+```ts
+export interface NamespaceOptions {
+  /**
+   * The domain name for the namespace, such as foo.com
+   */
+  name: string;
+
+  /**
+   * The type of CloudMap Namespace to create in your cluster
+   *
+   * @default PrivateDns
+   */
+  type?: cloudmap.NamespaceType;
+
+  /**
+   * The Amazon VPC that you want to associate the namespace with. Required for Private DNS namespaces
+   *
+   * @default VPC of the cluster for Private DNS Namespace, otherwise none
```

Comment:
Does it make sense to ever have a VPC that's not the Cluster's VPC? I think not, right? Might as well leave it out then.

Reply:
True -- technically a cluster can span more than one VPC but in the CDK we're strongly enforcing a 1:1 cluster:vpc relationship so that's fine.